### ISSUE TITLE
Traversals 

### ISSUE THREAD

**Issue by turnersr** [indradb/indradb#59]
Hello,

Is there a way to walk the graph given a source node without explicitly chaining together node and edge properties as a series of "outbound" and "inbound"? 

Maybe depth-first search with an edge type filter?

Thanks!

**Comment by ysimonson** [comment]
Not at the moment (and it is admittedly a bit wordy!)

**Comment by turnersr** [comment]
Yeah, no worries. I don't mind trying to implement it if it would be helpful.  Any recommendations on how to implement a more generic travel like doeth first search along an edge type using indradb primitives?

**Comment by ysimonson** [comment]
You could just make a helper function with what you put. Otherwise, if you're looking to get generally more terse code, consider the python bindings. See an example of it's use here: https://github.com/indradb/wikipedia-example

**Comment by turnersr** [comment]
Here is my first attempt. 

This code requires that the Vertex struct is hashable. I made a pull request that does this.  

What do you think? 

```Rust
extern crate common;
extern crate indradb;
extern crate serde_json;
extern crate uuid;
extern crate rand;


use serde_json::de::from_str;
use std::u32;
use uuid::Uuid;
use rand::Rng;

use std::collections::HashMap;

use indradb::{EdgeKey, SpecificEdgeQuery, Type, Vertex, VertexQuery};

// Traits!
use indradb::{Datastore, MemoryDatastore, MemoryTransaction, Transaction};

use indradb::{
    EdgeQueryExt, RangeVertexQuery, SpecificVertexQuery, VertexPropertyQuery, VertexQueryExt,
};

use std::hash::{Hash, Hasher};

#[derive(Debug, Clone)]
struct Node {
    id: u64,
}

#[derive(Debug, Clone)]
struct Edge {
    src: Node,
    dst: Node,
}

#[derive(Debug, Clone)]
struct Graph {
    edges: Vec<Edge>,
    nodes: Vec<Node>,
}

impl Graph {
    fn new(n: usize) -> Graph {
        let mut nodes = Vec::<Node>::with_capacity(n);
        for i in 1..n {
            nodes.push(Node { id: i as u64 });
        }
        let edges = Vec::new();

        return Graph {
            nodes: nodes,
            edges: edges,
        };
    }
}

fn uuid_create(i: i64) -> uuid::Uuid {
    let x = format!("{:032X}", i).to_string();
    let my_uuid = Uuid::parse_str(&x).unwrap();
    return my_uuid;
}

fn create_prufer_sequence(k: u64) -> Vec<u64> {
    let s = [4, 4, 4, 5];

let mut rng = rand::thread_rng();
let numbers: Vec<u64> = (0..k).map(|_| {
    rng.gen_range(1, 100)
}).collect();

    return numbers;
}

fn convert_prufer_to_tree(sequence: Vec<u64>) -> Graph {
    let n = sequence.len();
    let mut t = Graph::new((n + 3) as usize);
    let mut degree = HashMap::new();

    for i in t.nodes.iter() {
        degree.insert(i.id, 1);
    }

    for i in sequence.iter() {
        let d = degree.entry(*i).or_insert(1);
        *d += 1;
    }

    for i in sequence.iter() {
        for j in t.nodes.iter() {
            if degree[&j.id] == 1 {
                let e = Edge {
                    src: Node { id: *i },
                    dst: Node { id: j.id },
                };
                t.edges.push(e);
                let i_node = degree.entry(*i).or_default();
                *i_node -= 1;
                let j_node = degree.entry(j.id).or_default();
                *j_node -= 1;
                break;
            }
        }
    }

    let mut u = 0;
    let mut v = 0;

    for i in t.nodes.iter() {
        if degree[&i.id] == 1 {
            if u == 0 {
                u = i.id;
            } else {
                v = i.id;
                break;
            }
        }
    }

    let e = Edge {
        src: Node { id: u },
        dst: Node { id: v },
    };

    t.edges.push(e);

    let d_u = degree.entry(u).or_default();
    *d_u -= 1;
    let d_v = degree.entry(u).or_default();
    *d_v -= 1;

    return t;
}

fn convert_graph_to_db<T>(
    trans: &T,
    graph: Graph,
    node_type: &indradb::Type,
    edge_type: &indradb::Type,
) -> i64
where
    T: Transaction,
{
    for e in graph.edges.iter() {
        let src = &e.src.id;
        let dst = &e.dst.id;

        let outbound_v = Vertex::with_id(uuid_create(*src as i64), node_type.clone());
        let inbound_v = Vertex::with_id(uuid_create(*dst as i64), node_type.clone());

        trans.create_vertex(&outbound_v).unwrap();
        trans.create_vertex(&inbound_v).unwrap();

        let key = EdgeKey::new(outbound_v.id, edge_type.clone(), inbound_v.id);

        trans.create_edge(&key).unwrap();
    }

    return 0;
}

fn dfs_traversal<T> (trans : &T, start_vertex : indradb::Vertex, edge_type: &Type) -> HashMap<indradb::Vertex, u64>
where
    T: Transaction,
{

    let mut s = Vec::<indradb::Vertex>::new();

    s.push(start_vertex);

    let mut discovered_map: HashMap<indradb::Vertex, u64> = HashMap::new();
    let mut order = 0; 
    
    while !s.is_empty() {
        match s.pop() {
            Some(vertex) => {
                match discovered_map.get(&vertex) {
                    None => {

                        let query = SpecificVertexQuery::single(vertex.id)
                            .outbound(u32::MAX)
                            .t(edge_type.clone())
                            .inbound(u32::MAX);
                        
                        discovered_map.insert(vertex, order);

                        let e = trans.get_vertices(query.clone()).unwrap();

                        for v in e {
                            s.push(v);
                        }

                        order += 1;

                    }
                    Some(_) => {}
                }
            }
            None => {}
        }
    }

    return discovered_map;
}

fn main() {
    // Node Types
    let indra = MemoryDatastore::default();
    let trans = indra.transaction().expect("txn");

    let vertex_t = Type::new("test_vertex_type").unwrap();
    let edge_t = Type::new("test_edge_type").unwrap();

    let psequence = create_prufer_sequence(1000);
    let tree = convert_prufer_to_tree(psequence);

    //print!("{:x?}\n", tree);
    convert_graph_to_db(&trans, tree, &vertex_t, &edge_t);

    let query = SpecificVertexQuery::single(uuid_create(5))
        .outbound(u32::MAX)
        .t(edge_t)
        .inbound(u32::MAX);

    let e = trans.get_vertices(query.clone()).unwrap();

    println!("{:?}", e);


    let start_v = Vertex::with_id(uuid_create(5), vertex_t.clone());


    let traversal = dfs_traversal(&trans, start_v, &Type::new("test_edge_type").unwrap());

    print!("{:x?}", traversal);
   
}
```

**Comment by ysimonson** [comment]
That looks good! Alternatively, you could use the vertex UUID in `discovered_map`, in which case you don't need to worry about hashing the vertex (plus it would be faster.)
